File: src/main/java/org/taktik/icure/logic/InvoiceLogic.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic;

import java.util.List;
import java.util.Set;

import org.ektorp.ComplexKey;
import org.taktik.icure.db.PaginatedList;
import org.taktik.icure.db.PaginationOffset;
import org.taktik.icure.dto.data.LabelledOccurence;
import org.taktik.icure.entities.Invoice;
import org.taktik.icure.entities.embed.Delegation;
import org.taktik.icure.entities.embed.InvoiceType;
import org.taktik.icure.entities.embed.InvoicingCode;
import org.taktik.icure.entities.embed.MediumType;
import org.taktik.icure.exceptions.DeletionException;

public interface InvoiceLogic {
	Invoice createInvoice(Invoice invoice);
	String deleteInvoice(String invoiceId) throws DeletionException;
	Invoice getInvoice(String invoiceId);
	List<Invoice> getInvoices(List<String> strings);
	Invoice modifyInvoice(Invoice invoice);
	List<Invoice> updateInvoices(List<Invoice> invoices);

	Invoice addDelegation(String invoiceId, Delegation delegation);

	PaginatedList<Invoice> findByAuthor(String hcParty, String userId, Long fromDate, Long toDate, PaginationOffset<ComplexKey> paginationOffset);
	List<Invoice> listByHcPartyContacts(String hcParty, Set<String> contactIds);
	List<Invoice> listByHcPartyRecipientIds(String hcParty, Set<String> recipientIds);
	List<Invoice> listByHcPartyPatientSks(String hcParty, Set<String> patientSks);

	List<Invoice> listByHcPartySentMediumTypeInvoiceTypeSentDate(String hcParty, MediumType sentMediumType, InvoiceType invoiceType, boolean sent, Long fromDate, Long toDate);

	List<Invoice> listByHcPartySendingModeStatus(String hcParty, String sendingMode, String status, Long fromDate, Long toDate);

	List<Invoice> listByHcPartyGroupId(String hcParty, String groupId);

	List<Invoice> listByHcPartyRecipientIdsUnsent(String hcParty, Set<String> recipientIds);

	List<Invoice> listByHcPartyPatientSksUnsent(String hcParty, Set<String> secretPatientKeys);

	List<Invoice> listByServiceIds(Set<String> serviceIds);

	Invoice mergeInvoices(String hcParty, List<Invoice> invoices, Invoice destination) throws DeletionException;

	Invoice validateInvoice(String hcParty, Invoice invoice, String refScheme, String forceValue);

	List<Invoice> appendCodes(String hcParty, String userId, String insuranceId, Set<String> secretPatientKeys, InvoiceType type, MediumType sentMediumType, List<InvoicingCode> invoicingCodes, String invoiceId, Integer invoiceGraceTime);

	Invoice addDelegations(String invoiceId, List<Delegation> delegations);

	List<Invoice> removeCodes(String userId, Set<String> secretPatientKeys, String serviceId, List<String> tarificationIds);

	List<Invoice> listAllHcpsByStatus(String status, Long from, Long to, List<String> hcpIds);

	List<LabelledOccurence> getTarificationsCodesOccurences(String hcPartyId, long minOccurences);

	void solveConflicts();



}

File: src/main/kotlin/org/taktik/icure/asynclogic/InvoiceLogic.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */

package org.taktik.icure.asynclogic

import kotlinx.coroutines.flow.Flow
import org.taktik.couchdb.DocIdentifier
import org.taktik.couchdb.ViewQueryResultEvent
import org.taktik.icure.asyncdao.InvoiceDAO
import org.taktik.icure.db.PaginationOffset
import org.taktik.icure.domain.filter.chain.FilterChain
import org.taktik.icure.dto.data.LabelledOccurence
import org.taktik.icure.entities.Invoice
import org.taktik.icure.entities.embed.Delegation
import org.taktik.icure.entities.embed.InvoiceType
import org.taktik.icure.entities.embed.InvoicingCode
import org.taktik.icure.entities.embed.MediumType

interface InvoiceLogic : EntityPersister<Invoice, String> {
    suspend fun createInvoice(invoice: Invoice): Invoice?
    suspend fun deleteInvoice(invoiceId: String): DocIdentifier?

    suspend fun getInvoice(invoiceId: String): Invoice?
    fun getInvoices(ids: List<String>): Flow<Invoice>

    suspend fun modifyInvoice(invoice: Invoice): Invoice?
    fun modifyInvoices(invoices: List<Invoice>): Flow<Invoice>

    suspend fun addDelegation(invoiceId: String, delegation: Delegation): Invoice?
    fun findInvoicesByAuthor(hcPartyId: String, fromDate: Long?, toDate: Long?, paginationOffset: PaginationOffset<List<String>>): Flow<ViewQueryResultEvent>
    fun listInvoicesByHcPartyContacts(hcParty: String, contactIds: Set<String>): Flow<Invoice>
    fun listInvoicesByHcPartyAndRecipientIds(hcParty: String, recipientIds: Set<String?>): Flow<Invoice>
    fun listInvoicesByHcPartyAndPatientSks(hcParty: String, secretPatientKeys: Set<String>): Flow<Invoice>
    fun listInvoicesByHcPartySentMediumTypeInvoiceTypeSentDate(hcParty: String, sentMediumType: MediumType, invoiceType: InvoiceType, sent: Boolean, fromDate: Long?, toDate: Long?): Flow<Invoice>
    fun listInvoicesByHcPartySendingModeStatus(hcParty: String, sendingMode: String?, status: String?, fromDate: Long?, toDate: Long?): Flow<Invoice>
    fun listInvoicesByHcPartyAndGroupId(hcParty: String, inputGroupId: String): Flow<Invoice>
    fun listInvoicesByHcPartyAndRecipientIdsUnsent(hcParty: String, recipientIds: Set<String?>): Flow<Invoice>
    fun listInvoicesByHcPartyAndPatientSksUnsent(hcParty: String, secretPatientKeys: Set<String>): Flow<Invoice>
    fun listInvoicesByServiceIds(serviceIds: Set<String>): Flow<Invoice>

    suspend fun mergeInvoices(hcParty: String, invoices: List<Invoice>, destination: Invoice?): Invoice?

    suspend fun validateInvoice(hcParty: String, invoice: Invoice?, refScheme: String, forcedValue: String?): Invoice?
    fun appendCodes(hcPartyId: String, userId: String, insuranceId: String?, secretPatientKeys: Set<String>, type: InvoiceType, sentMediumType: MediumType, invoicingCodes: List<InvoicingCode>, invoiceId: String?, invoiceGraceTime: Int?): Flow<Invoice>

    suspend fun addDelegations(invoiceId: String, delegations: List<Delegation>): Invoice?
    fun removeCodes(userId: String, secretPatientKeys: Set<String>, serviceId: String, tarificationIds: List<String>): Flow<Invoice>
    fun listInvoicesHcpsByStatus(status: String, from: Long?, to: Long?, hcpIds: List<String>): Flow<Invoice>

    fun solveConflicts(): Flow<Invoice>

    suspend fun getTarificationsCodesOccurences(hcPartyId: String, minOccurences: Long): List<LabelledOccurence>
    fun listInvoicesIdsByTarificationsByCode(hcPartyId: String, codeCode: String, startValueDate: Long, endValueDate: Long): Flow<String>
    fun listInvoiceIdsByTarificationsByCode(hcPartyId: String, codeCode: String?, startValueDate: Long?, endValueDate: Long?): Flow<String>

    fun filter(filter: FilterChain<Invoice>): Flow<Invoice>
    fun getGenericDAO(): InvoiceDAO
}
